lib/casmutils/xtal/structure.cxx

```diff
@@ -1,46 +1,158 @@
+    }
+    else if (reference == "internal")
+    {
+        _update_basic_from_internals();
+        _update_simple_from_internals();
+    }
+    else
+    {
+        throw except::UserInputMangle("Check the update reference you used. Not compatible.");
+    }
+}
+
+void Structure::_update_simple_from_basic()
+{
+    this->casm_simplestructure = CASM::xtal::make_simple_structure(this->casm_basicstructure);
+}
+
+void Structure::_update_internals_from_basic()
+{
+    this->structure_lattice = casmutils::xtal::Lattice(this->casm_basicstructure.lattice());
+    std::vector<casmutils::xtal::Site> new_basis;
+    for (const auto& site : this->casm_basicstructure.basis())
+    {
+        new_basis.emplace_back(site, 0);
+    }
+    this->basis = new_basis;
+}
+
+void Structure::_update_basic_from_simple()
+{
+    /// Syncing Lattice
+    this->casm_basicstructure =
+        CASM::xtal::BasicStructure(CASM::xtal::Lattice(this->casm_simplestructure.lat_column_mat));
+    /// Lattice has been synced
+    auto& basic_basis = this->casm_basicstructure.set_basis();
+    for (int index = 0; index < this->casm_simplestructure.n_atom(); index++)
+    {
+        const auto& info = this->casm_simplestructure.info(CASM::xtal::SimpleStructure::SpeciesMode::ATOM);
+        Eigen::Vector3d raw_coord = info.coord(index);
+        basic_basis.emplace_back(CASM::xtal::Coordinate(raw_coord, this->casm_basicstructure.lattice(), CASM::CART),
+                                 info.names[index]);
+    }
+}
+
+void Structure::_update_internals_from_simple()
+{
+    this->structure_lattice = casmutils::xtal::Lattice(casm_simplestructure.lat_column_mat);
+    std::vector<casmutils::xtal::Site> new_basis;
+    for (int index = 0; index < this->casm_simplestructure.n_atom(); index++)
+    {
+        const auto& info = this->casm_simplestructure.info(CASM::xtal::SimpleStructure::SpeciesMode::ATOM);
+        Eigen::Vector3d raw_coord = info.coord(index);
+        new_basis.emplace_back(casmutils::xtal::Coordinate(raw_coord), info.names[index]);
+    }
+    this->basis = new_basis;
+}
+
+void Structure::_update_basic_from_internals()
+{
+    /// Syncing Lattice
+    this->casm_basicstructure =
+        CASM::xtal::BasicStructure(CASM::xtal::Lattice(this->structure_lattice.column_vector_matrix()));
+    /// Lattice has been synced
+    auto& basic_basis = this->casm_basicstructure.set_basis();
+    for (const auto& site : this->basis)
+    {
+        basic_basis.emplace_back(CASM::xtal::Coordinate(site.cart(), this->casm_basicstructure.lattice(), CASM::CART),
+                                 site.label());
+    }
+}
+
+void Structure::_update_simple_from_internals()
 {
-    throw except::NotImplemented();
-    // You are dealing with CASM::Array, but we want to return a std::vector
-    // std::vector<Site> basis(this->basis.begin(), this->basis.end());
-    // return basis;
+    /// This seems like a horrible way to do this but it's because SimpleStructure only has a
```

Comment:
Yes, it could probably be nicer but it'll be fine for now. We can hopefully get a better constructor in CASM eventually. Just add a little `TODO` note and move along.